File: test/language/offsets/cpp/TernaryOperatorOffsetTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#include "zserio/CppRuntimeException.h"

#include "offsets/ternary_operator_offset/TernaryOffset.h"

namespace offsets
{
namespace ternary_operator_offset
{

using allocator_type = TernaryOffset::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;

class TernaryOperatorOffsetTest : public ::testing::Test
{
protected:
    void writeTernaryOffsetToByteArray(zserio::BitStreamWriter& writer, bool isFirstOffsetUsed,
            bool writeWrongOffset)
    {
        writer.writeBool(isFirstOffsetUsed);
        if (isFirstOffsetUsed)
        {
            writer.writeBits((writeWrongOffset) ? WRONG_FIELD_OFFSET : FIELD_OFFSET, 32);
            writer.writeBits(WRONG_FIELD_OFFSET, 32);
        }
        else
        {
            writer.writeBits(WRONG_FIELD_OFFSET, 32);
            writer.writeBits((writeWrongOffset) ? WRONG_FIELD_OFFSET : FIELD_OFFSET, 32);
        }
        writer.writeSignedBits(FIELD_VALUE, 32);
    }

    void checkTernaryOffset(const TernaryOffset& ternaryOffset, bool isFirstOffsetUsed)
    {
        ASSERT_EQ(isFirstOffsetUsed, ternaryOffset.getIsFirstOffsetUsed());
        if (isFirstOffsetUsed)
        {
            ASSERT_EQ(FIELD_OFFSET, ternaryOffset.getOffsets()[0]);
            ASSERT_EQ(WRONG_FIELD_OFFSET, ternaryOffset.getOffsets()[1]);
        }
        else
        {
            ASSERT_EQ(WRONG_FIELD_OFFSET, ternaryOffset.getOffsets()[0]);
            ASSERT_EQ(FIELD_OFFSET, ternaryOffset.getOffsets()[1]);
        }
        ASSERT_EQ(FIELD_VALUE, ternaryOffset.getValue());
    }

    void fillTernaryOffset(TernaryOffset& ternaryOffset, bool isFirstOffsetUsed, bool createWrongOffset)
    {
        ternaryOffset.setIsFirstOffsetUsed(isFirstOffsetUsed);
        const vector_type<uint32_t> offsets = { WRONG_FIELD_OFFSET, WRONG_FIELD_OFFSET };
        ternaryOffset.setOffsets(offsets);
        ternaryOffset.setValue(FIELD_VALUE);
        if (!createWrongOffset)
            ternaryOffset.initializeOffsets();
    }

    void testOffset(bool isFirstOffsetUsed)
    {
        TernaryOffset ternaryOffset;
        const bool writeWrongOffset = false;
        fillTernaryOffset(ternaryOffset, isFirstOffsetUsed, writeWrongOffset);

        zserio::BitStreamWriter writer(bitBuffer);
        ternaryOffset.write(writer);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        const TernaryOffset readTernaryOffset(reader);
        checkTernaryOffset(readTernaryOffset, isFirstOffsetUsed);
    }

    void testOffsetWriteWrong(bool isFirstOffsetUsed)
    {
        TernaryOffset ternaryOffset;
        const bool writeWrongOffset = true;
        fillTernaryOffset(ternaryOffset, isFirstOffsetUsed, writeWrongOffset);

        zserio::BitStreamWriter writer(bitBuffer);
        EXPECT_THROW(ternaryOffset.write(writer), zserio::CppRuntimeException);
    }

    void testOffsetReadWrong(bool isFirstOffsetUsed)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        const bool writeWrongOffset = true;
        writeTernaryOffsetToByteArray(writer, isFirstOffsetUsed, writeWrongOffset);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        EXPECT_THROW(TernaryOffset ternaryOffset(reader), zserio::CppRuntimeException);
    }

    static const uint32_t WRONG_FIELD_OFFSET;
    static const uint32_t FIELD_OFFSET;
    static const int32_t FIELD_VALUE;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const uint32_t TernaryOperatorOffsetTest::WRONG_FIELD_OFFSET = 0;
const uint32_t TernaryOperatorOffsetTest::FIELD_OFFSET = (1 + 32 + 32 + /* align */ + 7) / 8;
const int32_t TernaryOperatorOffsetTest::FIELD_VALUE = 0xABCD;

TEST_F(TernaryOperatorOffsetTest, firstOffset)
{
    const bool isFirstOffsetUsed = true;
    testOffset(isFirstOffsetUsed);
}

TEST_F(TernaryOperatorOffsetTest, firstOffsetWriteWrong)
{
    const bool isFirstOffsetUsed = true;
    testOffsetWriteWrong(isFirstOffsetUsed);
}

TEST_F(TernaryOperatorOffsetTest, firstOffsetReadWrong)
{
    const bool isFirstOffsetUsed = true;
    testOffsetReadWrong(isFirstOffsetUsed);
}

TEST_F(TernaryOperatorOffsetTest, secondOffset)
{
    const bool isFirstOffsetUsed = false;
    testOffset(isFirstOffsetUsed);
}

TEST_F(TernaryOperatorOffsetTest, secondOffsetWriteWrong)
{
    const bool isFirstOffsetUsed = false;
    testOffsetWriteWrong(isFirstOffsetUsed);
}

TEST_F(TernaryOperatorOffsetTest, secondOffsetReadWrong)
{
    const bool isFirstOffsetUsed = false;
    testOffsetReadWrong(isFirstOffsetUsed);
}

} // namespace ternary_operator_offset
} // namespace offsets

File: test/language/offsets/python/TernaryOperatorOffsetTest.py
import unittest
import zserio

from testutils import getZserioApi

class TernaryOperatorOffsetTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "offsets.zs").ternary_operator_offset

    def testFirstOffset(self):
        isFirstOffsetUsed = True
        self._testOffset(isFirstOffsetUsed)

    def testFirstOffsetWriteWrong(self):
        isFirstOffsetUsed = True
        self._testOffsetWriteWrong(isFirstOffsetUsed)

    def testFirstOffsetReadWrong(self):
        isFirstOffsetUsed = True
        self._testOffsetReadWrong(isFirstOffsetUsed)

    def testSecondOffset(self):
        isFirstOffsetUsed = False
        self._testOffset(isFirstOffsetUsed)

    def testSecondOffsetWriteWrong(self):
        isFirstOffsetUsed = False
        self._testOffsetWriteWrong(isFirstOffsetUsed)

    def testSecondOffsetReadWrong(self):
        isFirstOffsetUsed = False
        self._testOffsetReadWrong(isFirstOffsetUsed)

    def _writeTernaryOffsetToStream(self, writer, isFirstOffsetUsed, writeWrongOffset):
        writer.write_bool(isFirstOffsetUsed)
        if isFirstOffsetUsed:
            writer.write_bits(self.WRONG_FIELD_OFFSET if writeWrongOffset else self.FIELD_OFFSET, 32)
            writer.write_bits(self.WRONG_FIELD_OFFSET, 32)
        else:
            writer.write_bits(self.WRONG_FIELD_OFFSET, 32)
            writer.write_bits(self.WRONG_FIELD_OFFSET if writeWrongOffset else self.FIELD_OFFSET, 32)
        writer.write_signed_bits(self.FIELD_VALUE, 32)

    def _checkTernaryOffset(self, ternaryOffset, isFirstOffsetUsed):
        self.assertEqual(isFirstOffsetUsed, ternaryOffset.is_first_offset_used)
        if isFirstOffsetUsed:
            self.assertEqual(self.FIELD_OFFSET, ternaryOffset.offsets[0])
            self.assertEqual(self.WRONG_FIELD_OFFSET, ternaryOffset.offsets[1])
        else:
            self.assertEqual(self.WRONG_FIELD_OFFSET, ternaryOffset.offsets[0])
            self.assertEqual(self.FIELD_OFFSET, ternaryOffset.offsets[1])

        self.assertEqual(self.FIELD_VALUE, ternaryOffset.value)

    def _createTernaryOffset(self, isFirstOffsetUsed, createWrongOffset):
        ternaryOffset = self.api.TernaryOffset(isFirstOffsetUsed,
                                               [self.WRONG_FIELD_OFFSET, self.WRONG_FIELD_OFFSET],
                                               self.FIELD_VALUE)
        if not createWrongOffset:
            ternaryOffset.initialize_offsets()

        return ternaryOffset

    def _testOffset(self, isFirstOffsetUsed):
        writeWrongOffset = False
        ternaryOffset = self._createTernaryOffset(isFirstOffsetUsed, writeWrongOffset)

        writer = zserio.BitStreamWriter()
        ternaryOffset.write(writer)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readTernaryOffset = self.api.TernaryOffset.from_reader(reader)

        self._checkTernaryOffset(readTernaryOffset, isFirstOffsetUsed)

    def _testOffsetWriteWrong(self, isFirstOffsetUsed):
        writeWrongOffset = True
        ternaryOffset = self._createTernaryOffset(isFirstOffsetUsed, writeWrongOffset)

        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            ternaryOffset.write(writer)

    def _testOffsetReadWrong(self, isFirstOffsetUsed):
        writer = zserio.BitStreamWriter()
        writeWrongOffset = True
        self._writeTernaryOffsetToStream(writer, isFirstOffsetUsed, writeWrongOffset)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        with self.assertRaises(zserio.PythonRuntimeException):
            self.api.TernaryOffset.from_reader(reader)

    WRONG_FIELD_OFFSET = 0
    FIELD_OFFSET = (1 + 32 + 32 + 7) / 8
    FIELD_VALUE = 0xABCD
